### ISSUE TITLE
How are the gradients computed?

### ISSUE THREAD

**Issue by SSamDav** [lvwerra/trl#73]
Foremost, happy new year to all!! 🎆 

Congrats for the nice repo that you have!

I have some questions about how the gradients are computed?

So in the notebooks you start by generate continuations 
```
 response = gpt2_model.generate(query_tensors[i].unsqueeze(dim=0),
                                       max_new_tokens=gen_len, **gen_kwargs)
```

that because of the generate, generate the response without gradients. Then inside the `step()` you do:
```
 with torch.no_grad():
    logits, _, v = self.model(input_ids)
    ref_logits, _, _ = self.ref_model(input_ids)

logprobs = logprobs_from_logits(logits[:, :-1, :], input_ids[:, 1:])
ref_logprobs = logprobs_from_logits(ref_logits[:, :-1, :], input_ids[:, 1:])
```
that is used to compute the loss that is then back propagated through the model. 
So I don't see where in the code you run the forward with the gradients, Am I missing something? 

Thanks for the help!

**Comment by SSamDav** [comment]
I think I got the answer, It is inside the loss function 😅
